File: new-resource-detector/src/test/java/nl/haploid/octowight/service/ResourceDetectorsServiceTest.java
package nl.haploid.octowight.service;

import mockit.Injectable;
import mockit.Mocked;
import mockit.StrictExpectations;
import mockit.Tested;
import nl.haploid.octowight.AtomChangeEvent;
import nl.haploid.octowight.AtomGroup;
import nl.haploid.octowight.TestData;
import nl.haploid.octowight.detector.ResourceDetector;
import nl.haploid.octowight.registry.data.ResourceRoot;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;

public class ResourceDetectorsServiceTest {

	@Tested
	private ResourceDetectorsService service;

	@Injectable
	private List<ResourceDetector> mockDetectors;

	@Test
	public void testGetDetectorsForAtomType(final @Mocked ResourceDetector mockDetector1,
	                                        final @Mocked ResourceDetector mockDetector2) {
		final List<ResourceDetector> expectedDetectors = Collections.singletonList(mockDetector1);
		final String atomType = "sterling";
		new StrictExpectations(service) {{
			service.getDetectors();
			times = 1;
			result = Arrays.asList(mockDetector1, mockDetector2);
			mockDetector1.getAtomTypes();
			times = 1;
			result = Collections.singletonList(atomType);
			mockDetector2.getAtomTypes();
			times = 1;
			result = Collections.singletonList("cooper");
		}};
		final List<ResourceDetector> actualDetectors = service.getDetectorsForAtomType(atomType);
		assertEquals(expectedDetectors, actualDetectors);
	}

	@Test
	public void testDetectResources(final @Mocked ResourceDetector mockDetector) {
		final AtomChangeEvent event1 = TestData.atomChangeEvent("draper");
		final AtomChangeEvent event2 = TestData.atomChangeEvent("pryce");
		final AtomGroup atomGroup = new AtomGroup();
		atomGroup.setAtomOrigin(event1.getAtomOrigin());
		atomGroup.setAtomType(event1.getAtomType());
		final List<AtomChangeEvent> events = Arrays.asList(event1, event2);
		final List<ResourceRoot> expectedResourceRoots = Collections.singletonList(TestData.resourceRoot(null));
		new StrictExpectations(service) {{
			service.getDetectorsForAtomType("draper");
			times = 1;
			result = Collections.singletonList(mockDetector);
			mockDetector.detect(events);
			times = 1;
			result = expectedResourceRoots;
		}};
		final List<ResourceRoot> actualResourceRoots = service.detectResources(atomGroup, events);
		assertEquals(expectedResourceRoots, actualResourceRoots);
	}
}

File: new-resource-detector/src/test/scala/nl/haploid/octowight/service/ResourceDetectorsServiceTest.scala
package nl.haploid.octowight.service

import nl.haploid.octowight._
import nl.haploid.octowight.detector.ResourceDetector
import org.easymock.EasyMock

class ResourceDetectorsServiceTest extends AbstractTest {

  behavior of "Resource detectors service"

  it should "get detectors for atom category" in {
    val resourceDetectorsService = EasyMock.createMockBuilder(classOf[ResourceDetectorsService])
      .addMockedMethod("detectors")
      .createMock()
    val detector1 = mock[ResourceDetector]
    val detector2 = mock[ResourceDetector]
    val expectedDetectors = List(detector1)
    val atomCategory = "sterling"
    expecting {
      resourceDetectorsService.detectors andReturn List(detector1, detector2) once()
      detector1.atomCategories andReturn List(atomCategory) once()
      detector2.atomCategories andReturn List("cooper") once()
    }
    whenExecuting(resourceDetectorsService, detector1, detector2) {
      val actualDetectors = resourceDetectorsService.detectorsForAtomCategory(atomCategory)
      actualDetectors should be(expectedDetectors)
    }
  }

  it should "detect resources" in {
    val resourceDetectorsService = EasyMock.createMockBuilder(classOf[ResourceDetectorsService])
      .addMockedMethod("detectorsForAtomCategory", classOf[String])
      .createMock()
    val detector = mock[ResourceDetector]
    val event1 = TestData.atomChangeEvent("draper")
    val event2 = TestData.atomChangeEvent("pryce")
    val atomGroup = new AtomGroup(origin = event1.atomOrigin, category = event1.atomCategory)
    val events = List(event1, event2)
    val expectedResourceRoots = List(TestData.resourceRoot(null))
    expecting {
      resourceDetectorsService.detectorsForAtomCategory("draper") andReturn List(detector) once()
      detector.detect(events) andReturn expectedResourceRoots once()
    }
    whenExecuting(resourceDetectorsService, detector) {
      val actualResourceRoots = resourceDetectorsService.detectResources(atomGroup, events)
      actualResourceRoots should be(expectedResourceRoots)
    }
  }
}
